pkg/maps/fragmap/fragmap.go

```diff
@@ -0,0 +1,109 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fragmap
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "map-events")
+)
+
+const (
+	// MapName is the name of the map used to retrieve L4 ports associated
+	// to the datagram to which an IPv4 belongs.
+	MapName = "cilium_ipv4_frag_datagrams"
+	// MaxEntries is the number of entries in the LRU map
+	MaxEntries = 8192
+)
+
+// FragmentKey must match 'struct ipv4_frag_id' in "bpf/lib/ipv4.h".
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
+type FragmentKey struct {
+	destAddr   types.IPv4 `align:"daddr"`
+	sourceAddr types.IPv4 `align:"saddr"`
+	id         uint16     `align:"id"`
+	proto      uint8      `align:"proto"`
+}
+
+// FragmentValue must match 'struct ipv4_frag_l4ports' in "bpf/lib/ipv4.h".
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type FragmentValue struct {
+	destPort   uint16 `align:"dport"`
+	sourcePort uint16 `align:"sport"`
+}
+
+// GetKeyPtr returns the unsafe pointer to the BPF key.
+func (k *FragmentKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *FragmentValue) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// String converts the key into a human readable string format.
+func (k *FragmentKey) String() string {
+	return fmt.Sprintf("%s --> %s, %d, %d", k.sourceAddr, k.destAddr, k.proto, k.id)
+}
+
+// String converts the value into a human readable string format.
+func (v *FragmentValue) String() string {
+	return fmt.Sprintf("%d, %d", v.destPort, v.sourcePort)
+}
+
+// NewValue returns a new empty instance of the structure representing the BPF
+// map value.
+func (k FragmentKey) NewValue() bpf.MapValue { return &FragmentValue{} }
+
+// InitMap creates the signal map in the kernel.
+func InitMap() error {
+	if !option.Config.EnableIPv4 || !option.Config.EnableIPv4FragmentsTracking {
+		return nil
+	}
+
+	pm := probes.NewProbeManager()
+	supportedMapTypes := pm.GetMapTypes()
+	if !supportedMapTypes.HaveLruHashMapType {
+		return nil
+	}
+
+	fragMap := bpf.NewMap(MapName,
+		bpf.MapTypeLRUHash,
+		&FragmentKey{},
+		int(unsafe.Sizeof(FragmentKey{})),
+		&FragmentValue{},
+		int(unsafe.Sizeof(FragmentValue{})),
+		MaxEntries,
+		0,
+		0,
+		bpf.ConvertKeyValue,
+	)
+	if _, err := fragMap.OpenOrCreate(); err != nil {
+		return err
+	}
+	err := fragMap.Close()
+	return err
+}
+
+// +k8s:deepcopy-gen=true
```

Comment:
huh, I always thought that this statement is supposed to appear above a structure. /cc @aanm how does this actually work?

Reply:
(overall question, why here at the end of the file and is there a specific purpose?)

Reply:
Seems to be a mistake, I think I moved some struct upwards in the file and forgot to move that line (and added it again to the struct). I'll clean it up, thanks for spotting that!